wen-restart/src/last_voted_fork_slots_aggregate.rs

```diff
@@ -90,80 +125,131 @@ impl LastVotedForkSlotsAggregate {
+            if let Some(sender_stake) = Self::validator_stake_at_slot(&self.root_bank, slot, from) {
+                *entry = entry.saturating_sub(sender_stake);
+                let repair_threshold_stake = (Self::total_stake_at_slot(&self.root_bank, slot)
+                    .unwrap() as f64
+                    * self.repair_threshold) as u64;
+                if *entry < repair_threshold_stake {
+                    self.slots_to_repair.remove(slot);
+                }
             }
         }
         for slot in new_slots_set.difference(&old_slots_set) {
             let entry = self.slots_stake_map.entry(*slot).or_insert(0);
-            *entry = entry.saturating_add(sender_stake);
-            if *entry >= threshold_stake {
-                self.slots_to_repair.insert(*slot);
+            if let Some(sender_stake) = Self::validator_stake_at_slot(&self.root_bank, slot, from) {
+                *entry = entry.saturating_add(sender_stake);
+                let repair_threshold_stake = (Self::total_stake_at_slot(&self.root_bank, slot)
+                    .unwrap() as f64
+                    * self.repair_threshold) as u64;
+                if *entry >= repair_threshold_stake {
+                    self.slots_to_repair.insert(*slot);
+                }
             }
         }
-        Some(record)
+        true
     }
 
-    pub(crate) fn active_percent(&self) -> f64 {
-        let total_stake = self.epoch_stakes.total_stake();
-        let total_active_stake = self.active_peers.iter().fold(0, |sum: u64, pubkey| {
-            sum.saturating_add(Self::validator_stake(&self.epoch_stakes, pubkey))
-        });
-        total_active_stake as f64 / total_stake as f64 * 100.0
+    fn update_epoch_info_vec(&mut self) {
+        let highest_repair_slot = self.slots_to_repair.last();
+        match highest_repair_slot {
+            Some(slot) => {
+                let current_epoch = self.root_bank.epoch_schedule().get_epoch(*slot);
+                let last_entry_epoch = self.epoch_info_vec.last().unwrap().epoch;
+                match last_entry_epoch.cmp(&current_epoch) {
+                    Ordering::Less => {
+                        // wandering into new epoch, add to the end of state vec, we should have at most 2 entries.
+                        self.epoch_info_vec.push(LastVotedForkSlotsEpochInfo {
+                            epoch: current_epoch,
+                            total_stake: Self::total_stake_at_slot(&self.root_bank, slot).unwrap(),
+                            total_active_stake: 0,
+                        });
+                    }
+                    Ordering::Greater => {
+                        // Somehow wandering into new epoch and no more because someone changed votes, let's remove
+                        // the new epoch and update the old one.
+                        assert!(self.epoch_info_vec.first().unwrap().epoch == current_epoch);
+                        self.epoch_info_vec.truncate(1);
```

Comment:
This seems impossible given `slots_to_repair` is a btree and thus `let highest_repair_slot = self.slots_to_repair.last();` is monotonically increasing, and the epoch must always be greater. I think we can panic here

Reply:
It is possible, because I do `self.slots_to_repair.remove(slot);` above.

The logic is like this: we allow any user to replace the `RestartLastVotedForkSlots` they sent out (maybe they screwed up the ledger directory and tried again). So if some validator `X` sent out wrong Gossip messages last time claiming it voted up to slot `1000`, but now replace it with only slot `500`, then we have to subtract its stake from slots `500`~`1000`, this may lead the highest `slots_to_repair` to go backwards.

Reply:
this would be a good comment to add to that case

Reply:
I'm thinking now we should enforce for each validator that the `RestartLastVotedForkSlots` cannot go down, only monotonically increase.

There should be no reason somebody posts a lower last voted forks other than corruption/malice. 

In gossip we can enforce we take the latest one instead of latest one by timestamp. 

Also would simplify this code

Reply:
Yeah I thought about that at the beginning. But I was afraid that honest validator operators can't resolve the mistakes they had. Let's say someone has hot standby set up, one of them has a bad state voting into the far future, the other one is good. The operator made the mistake to restart the bad one first, then whatever he does, he can't revert back to the good one any more.

Reply:
Per our conversation, will send out another PR to forbid any update to RestartLastVotedForkSlots/HeaviestFork. Because that has little to do with epoch_stakes.